### ISSUE TITLE
Endless loop in uart pipe

### ISSUE THREAD

**Issue by erwango** [zephyrproject-rtos/zephyr#5298]
Reported in devel mailing list by Dmitri Shmidt.

"The problem is in uart_pipe_isr() logic. It is never-ending (spinning)
after first time entrance: it is still processing RX but its nature doesn't allow any user
threads to get control."

https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-December/008533.html

**Comment by pfalcon** [comment]
Note that I saw similar effects in https://github.com/zephyrproject-rtos/zephyr/pull/42, and while working on irq-driven console in general: https://github.com/zephyrproject-rtos/zephyr/pull/49.

**Comment by erwango** [comment]
> while working on irq-driven console in general: #49.

Wouldn't it be because there is always a tx associated to rx in console case?
Hence TXE is set to 0 and while loop is exited.
I have been to reproduce issue by commenting out insert_char() in  uart_console_isr, but no issue as is.

**Comment by pfalcon** [comment]
Sure, for the console case it was solved by following invariants later documented in https://github.com/zephyrproject-rtos/zephyr/pull/159 , https://github.com/zephyrproject-rtos/zephyr/pull/281. And I had to patch existing test in #42 to follow them too.

So, my points is that such problems were known, and there's an existing pattern how to resolve them. And my concern that this case wasn't resolved in a way to break previous cases. (I'm going to look at #5299 in detail, from the title, it sounds good).
